Rubberduck.VBEEditor/Events/VBENativeServices.cs

```diff
@@ -57,136 +40,94 @@ public static void UnhookEvents()
             lock (ThreadLock)
             {
                 User32.UnhookWinEvent(_eventHandle);
-                foreach (var info in TrackedWindows.Values)
-                {
-                    info.Subclass.FocusChange -= FocusDispatcher;
-                    info.Subclass.Dispose();
-                }
+                Subclasses.Dispose();
                 VBEEvents.Terminate();
+                _vbe.Dispose();
                 _vbe = null;
             }
         }
 
+        private static void Attach(IntPtr hwnd)
+        {
+            var subclass = Subclasses.Subclass(hwnd);
+
+            if (subclass == null)
+            {
+                return;
+            }
+
+            if (subclass is IFocusProvider focusSource)
+            {
+                focusSource.FocusChange += FocusDispatcher;
+            }
+
+            if (subclass is IWindowEventProvider keyboardListener)
+            {
+                keyboardListener.KeyDown += KeyDownDispatcher;
+            }
+        }
+
         public static void VbeEventCallback(IntPtr hWinEventHook, uint eventType, IntPtr hwnd, int idObject, int idChild,
             uint dwEventThread, uint dwmsEventTime)
         {
-            if (hwnd == IntPtr.Zero) { return; }
-            //This is an output window firehose, leave this here, but comment it out when done.
-            //if (idObject != (int)ObjId.Cursor) { Debug.WriteLine("Hwnd: {0:X4} - EventType {1:X4}, idObject {2}, idChild {3}", (int)hwnd, eventType, idObject, idChild); }
+            if (hwnd == IntPtr.Zero || _vbe.IsWrappingNullReference) { return; }
 
+#if THIRSTY_DUCK && DEBUG
+            //This is an output window firehose, viewer discretion is advised.
+            if (idObject != (int)ObjId.Cursor) { Debug.WriteLine("Hwnd: {0:X4} - EventType {1:X4}, idObject {2}, idChild {3}", (int)hwnd, eventType, idObject, idChild); }
+#endif
             var windowType = hwnd.ToWindowType();
-            var nameBuilder = new StringBuilder(255);
-            User32.GetClassName(hwnd, nameBuilder, 255);
-            var className = nameBuilder.ToString();
 
-            if (className == "NameListWndClass")
+            if (windowType == WindowType.IntelliSense)
```

Comment:
I wonder if it is worthwhile to broadcast a event rather than maintaining a long list of `if`/`else if`s for unrelated classes. The `Subclassable` could be required to implement a delegate which would be then added to a list, and thus enumerated instead.

Reply:
The `VbeEventCallback` it there for handling application (VBE) level events that aren't tied to specific windows. The more important parameters are the `eventType` and `idObject` - I only grab the `windowType` before the `if`/`else if`s because it's used in multiple tests and it involves an API call.

The IntelliSense window isn't subclassed (I'm assuming this event filter is for autocomplete), and it wouldn't really be practical to do so. I haven't really explored it because I'd just assumed it was being continually recycled. 

The only other test performed here that is related to a specific window that can be assumed to be subclassed is the caret change notification. The reason that one is being performed at the VBE level is because it doesn't show up as cleanly in the individual windows' message pumps. About the only way to pick that out cleanly in the `CodePaneSubclass` is to listen for the "blinking" of the caret and then check its position against the last known caret position. That would be much more complex and unreliable.